text/0002-new-architecture.md

```diff
@@ -0,0 +1,243 @@
+a good enough system for running at scale and with some modifications this might also work well enough
+for single organization installations.
+
+### Move Sourcemap Processing to Symbolicator
+
+Our sourcemap processing system is running in a mix of Rust and Python code by fetching a bunch of data
+via HTTP and from the database models to then send the data fetched into a Rust module for resolving.
+This is generally quite inefficient but we are also running into limitations in the workers.  For instance
+we rely exclusively on memcache for caches which means that we are limited by our cache size limitations.
+Anything above the cache size is not processable which can cause a bad user experience for users depending
+on very large source maps.
+
+Symbolicator has a file system based caching system with GCS backing for rapid synching of new symbolicator
+instances we could leverage.  Additionally symbolicator has a better HTTP fetching system than Sentry as
+it's able to concurrencly fetch whereas the Sentry codebase is not.
+
+Future feature improvements on the source map side might also demand more complexity in the processing side
+which we are not currently considering given already existing scalability concerns on the existing workers.
+
+## Python Monolith
+
+These are changes to break up the monolith.  The goals here are largely that different teams can work largely
+uninterrupted of each other, even on the monolith.  For instance in an ideal scenario changes to an integration
+platform code do not necessarily have to run the entire Sentry testsuite on every commit.  Likewise UI code
+ideally does not build the entirety of the Sentry UI codebase.
+
+### Import Compartmentalization
+
+As a first start imports in the Sentry codebase should be compartmentalized as much as possible.  Today we have
+some catch all modules like `sentry.models` which import everything.  This means that it's hard for as to track
+the actual dependencies from different pieces of code and also hides away circular imports.  One of the results of
+this is that if you run a test against a single pytest module, you still pull in the entire codebase which already
+takes a couple of seconds.  It also means that it's harder for test infrastructure code to analyze the minimal set
+of dependencies that might need testing.
+
+Likewise code in `sentry.utils` should most likely no longer import models etc.  Some of that code might event
+move out of `sentry` entirely.
+
+### Move Code from Getsentry to Senty
+
+Currently quite a bit of code lives in `getsentry` for reasons of assumed convenience.  However this has created
+the situation that it is quite easy to ship regressions because the changes in `getsentry` were not considered.
+There is likely a whole range of code that does not need to live in `getsentry` and moving it to `sentry` would
+make the testsuite more reliable, faster to run and catch more regressions early.
```

Comment:
I could see usage periods and usage collection being something that could move to sentry. We would keep invoicing and subscriptions in getsentry but the usage subsystem could be moved out. This could also help align product features like orgstats and dynamic sampling suggestions with the usage periods that we use for billing in saas more easily.